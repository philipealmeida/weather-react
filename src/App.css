/* weather app styles */
@import url("https://fonts.googleapis.com/css2?family=Montserrat:wght@200;400;500;600&display=swap");

:root {
	margin: 0;
	padding: 0;
	--color-dark-blue: #100f3b;
	--color-mostard: #eec300;
	--color-gray-light: #f2fbff;
	--border-radius-primary: 1.63rem;
	--border-radius-b1: 0.5rem;
	--padding-wrapper: 3rem;
	--letter-spacing: -0.025em;
	--letter-spacing-2: -0.05em;
	--font-size-p17: 1.066rem;
	--font-size-p12: 0.75rem;
	--font-size-p10: 0.65rem;
}

@media (prefers-reduced-motion: no-preference) {}

body {
	margin: 0;
	padding: 0;
	min-height: 100vh;
	background: radial-gradient(50% 50% at 50% 50%, #a7cffd 0%, #7fb5f3 100%);
	display: flex;
	align-items: center;
	justify-content: center;
	justify-items: center;
	font-family: "Montserrat", sans-serif;
	letter-spacing: var(--letter-spacing);
	font-weight: 400;
}

.bold {
	font-weight: 600;
}

article {
	width: 100%;
	min-width: 1100px;
	background: radial-gradient(29.63% 90.27% at 32.01% 49.35%, #f9fdff 0%, #c9edfd 100%);
	border-radius: var(--border-radius-primary);
	width: 80vw;
	max-width: 1070px;
	box-shadow: 3px 5px 16px rgba(0, 0, 0, 0.18);
	display: flex;
	justify-content: space-between;
}

.hero {
	padding: var(--padding-wrapper);
	width: 100%;
}

.hero .hero-title {
	font-weight: 400;
	font-size: 1.93rem;
	letter-spacing: var(--letter-spacing-2);
	margin-top: 1.3rem;
	margin-bottom: 0;
}

.hero .filters button,
.hero .filters input {
	font-family: "Montserrat", sans-serif;
}

.hero .filters input {
	padding: 1.3rem 1rem;
	border-radius: var(--border-radius-b1);
	border: 0;
	width: 39%;
	font-size: 0.75rem;
	background: #fff url(/images/filter.svg) no-repeat 1rem center;
	padding-left: 3rem;
}

.hero .filters button {
	background: none;
	border: 0;
	padding: 1rem;
	padding-left: 0;
	margin-left: 1.4rem;
	font-weight: 500;
	cursor: pointer;
	color: var(--color-dark-blue);
}

.hero .cards .row {
	display: flex;
	flex-flow: row wrap;
	align-items: self-end;
}

.hero .cards .row:nth-of-type(2) {
	margin-top: 2rem;
}

.hero .cards .card:hover{
	transform: translateY(-5px);
}

.hero .cards .card {
	text-align: center;
	width: 21%;
	cursor: pointer;
	margin-left: 5%;
	margin-top: 3.4em;
	transition: 250ms ease;
}

.hero .cards .card:nth-child(4n+1){
	margin-left: 0;
}

.hero .cards .card .wrapper-image-card {
	width: 100%;
	border-radius: var(--border-radius-primary);
	overflow: hidden;
	transition: 250ms ease;
	max-height: 190px;
	aspect-ratio: 1/1.1;
}

.hero .cards .card .wrapper-image-card img {
	transition: 250ms ease;
	aspect-ratio: 1/1.1;
	object-fit: cover;
	width: 100%;
}

.hero .cards .card.active {
	filter: drop-shadow(2px 4px 4px rgba(0, 0, 0, 0.2));
}

.hero .cards .card.active .wrapper-image-card {
	aspect-ratio: 1/1.3;
}

.hero .cards .card.active .wrapper-image-card img {
	aspect-ratio: 1/1.3;
}

.hero .cards .card .card-title {
	margin: 0.45rem 0 0;
	font-size: var(--font-size-p10);
}

aside {
	display: flex;
	flex-flow: column;
	justify-content: space-between;
	min-width: 30%;
	padding: var(--padding-wrapper);
	background-color: var(--color-dark-blue);
	color: #fff;
	border-top-right-radius: var(--border-radius-primary);
	border-bottom-right-radius: var(--border-radius-primary);
}

aside header {
	display: flex;
	flex-flow: nowrap;
	align-items: center;
	justify-content: space-between;
	font-weight: 400;
}

aside header ul {
	align-items: center;
}

aside header ul,
aside header li {
	margin: 0;
	padding: 0;
}

aside header ul {
	list-style-type: none;
	display: flex;
	color: #fff;
}

aside header li {
	margin-right: 1rem;
}

aside header .alerts {
	display: inline-block;
	background-color: red;
	padding: 0.18rem 0.35rem;
	margin-left: 0.1em;
	font-size: 0.8rem;
	vertical-align: middle;
	border-radius: 6px;
	text-align: center;
}

aside header a {
	color: #fff;
	font-size: 0.7rem;
	text-decoration: none;
}

aside header .user-avatar {
	width: 42px;
	height: 42px;
	background: #fff url(/images/av-fig.jpg) no-repeat center center/cover;
	border-radius: 50%;
	position: relative;
}

aside .user-avatar:before {
	content: '';
	position: absolute;
	top: 0;
	width: 10px;
	height: 10px;
	background-color: red;
	right: 0;
	border-radius: 50%;
}

aside .daily-conditions {
	margin: 4.4rem 0;
	text-align: center;
	display: flex;
	flex-flow: column;
}

aside .daily-conditions .date-conditions {
	display: flex;
	align-items: center;
	justify-content: center;
}

aside .daily-conditions .date-conditions picture {
	width: 45px;
	margin-right: .5rem;
}

aside .daily-conditions .date-conditions picture img {
	width: 100%;
}

aside .daily-conditions .date-conditions .date-info {
	display: flex;
	text-align: left;
	flex-flow: column;
}

aside .daily-conditions .date-conditions .date-info .day {
	font-size: var(--font-size-p17);
	margin-bottom: 0.2rem;
}

aside .daily-conditions .date-conditions .date-info .week-day {
	font-size: var(--font-size-p10);
	font-weight: 200;
	color: var(--color-gray-light);
}

aside .daily-conditions .temperature-value {
	font-size: 6rem;
	font-weight: 200;
	vertical-align: top;
	line-height: 1;
	position: relative;
}

aside .daily-conditions .temperature-value .celsius {
	font-size: 2rem;
	font-size: 2rem;
	position: absolute;
	top: 0.5rem;
	color: var(--color-gray-light);
}

aside .daily-conditions .city-name {
	font-size: var(--font-size-p12);
	color: var(--color-gray-light);
}

aside .weekly-forecast {
	font-size: var(--font-size-p12);
	font-weight: 400;
	display: flex;
	flex-flow: column;
}

aside .weekly-forecast .title {
	font-size: 0.8rem;
}

aside .weekly-forecast .row {
	display: flex;
	margin-top: 1.5rem;
	justify-content: space-between;
	padding: 0 1.8em;
	align-content: center;
	text-align: center;
	font-size: 0.8rem;
}

aside .weekly-forecast .col div{
	margin-top: .5rem;
}

aside .weekly-forecast .col .min-temp,
aside .weekly-forecast .col .max-temp{
	font-size: 1rem;
}

aside .weekly-forecast picture{
	width: 40px;
	height: 40px;
	display: inline-block;
}

aside .weekly-forecast picture img{
	max-width: 100%;
}
