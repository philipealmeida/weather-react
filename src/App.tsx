import React, { FormEvent, useEffect, useState } from "react";
import { cities, weatherAPI } from './shared/utils/'
import { City, OpenWeather } from "./shared/types";
import './App.css'
import {Cards} from "./components/cards/Cards";
import {AsideHeader} from "./components/headers/AsideHeader";
import {Forecast} from "./components/forecast/Forecast";

function App() {
  const [filter, setFilter] = useState<string>('');
  const [filtered, setFiltered] = useState<City[]>(cities);
  const [data, setData] = useState<OpenWeather | null>(null);
  const [selectedCity, setSelectedCity] = useState<City>(cities[0]);

  const onFilterChange = (event: FormEvent<HTMLInputElement>) => {
    setFilter(event.currentTarget.value);
  };

  const resetFilter = () => {
    setFilter('');
  };

  const selectCity = (event: React.SyntheticEvent<Element, Event>, data: City) => {
    setSelectedCity(data);
  }

  useEffect(() => {
    const fetchData = async () => {
      try {
        const response = await fetch(`${weatherAPI.OPEN_WEATHER_URL}/forecast?lat=${selectedCity.lat}&lon=${selectedCity.long}&units=metric&lang=en&APPID=${weatherAPI.OPEN_WEATHER_API_KEY}`);
        const data = await response.json();
        setData(data);
      } catch (error) {
        throw new Error('Error on fetch OpenWeather API');
      }
    };
    fetchData();
  }, [selectedCity])

  useEffect(() => {
    const filtered = () => {
      const cities_filtered = cities.filter(c => c.name.toLowerCase().includes(filter.toLocaleLowerCase()) || filter === '');
      setFiltered(cities_filtered.length ? cities_filtered : cities);
    };
    filtered();
  }, [filter]);

  if (data) {
    return (
      <main className="App">
        <article>
          <section className="hero">
            <header>
              <section className="filters">
                <div className="wrapper-filter">
                  <input type="text" placeholder="Search City" value={filter} onChange={onFilterChange} />
                  <button type="button" onClick={resetFilter}>All places</button>
                </div>
              </section>
              <h1 className="hero-title">World<span className="bold"> Weather</span></h1>
            </header>
            <Cards cities={filtered} selectCity={selectCity} selectedCity={selectedCity} />
          </section>

          <aside>
            <AsideHeader/>
            <Forecast selectedCity={selectedCity} data={data}/>
          </aside>
        </article>
      </main>
    )
  }

  return <div></div>;
}

export default App;
