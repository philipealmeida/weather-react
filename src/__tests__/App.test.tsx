import { render, screen } from '@testing-library/react';

import App from '../App';
Object.defineProperty(window, 'matchMedia', {
    writable: true,
    value: jest.fn().mockImplementation(query => ({
      matches: false,
      media: query,
      onchange: null,
      addListener: jest.fn(), // Deprecated
      removeListener: jest.fn(), // Deprecated
      addEventListener: jest.fn(),
      removeEventListener: jest.fn(),
      dispatchEvent: jest.fn(),
    })),
  });
test('Renders main page correctly', async () => {
    const {container} = render(<App />);

    expect(container).toMatchSnapshot();
});