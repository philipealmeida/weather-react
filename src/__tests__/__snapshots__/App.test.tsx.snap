// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders main page correctly 1`] = `
<div>
  <main
    class="App"
  >
    <article>
      <section
        class="hero"
      >
        <header>
          <section
            class="filters"
          >
            <div
              class="wrapper-filter"
            >
              <input
                placeholder="Search City"
                type="text"
                value=""
              />
              <button
                type="button"
              >
                All places
              </button>
            </div>
          </section>
          <h1
            class="hero-title"
          >
            World 
            <span
              class="bold"
            >
              Weather
            </span>
          </h1>
        </header>
        <section
          class="cards"
        >
          <div
            class="row"
          >
            <div
              class="card active"
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Berlin"
                  src="/images/berlin.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Berlin
                , 
                Germany
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Paris"
                  src="/images/paris.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Paris
                , 
                France
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Porto"
                  src="/images/porto.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Porto
                , 
                Portugal
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="London"
                  src="/images/london.webp"
                />
              </div>
              <p
                class="card-title"
              >
                London
                , 
                England
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Tokyo"
                  src="/images/tokyo.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Tokyo
                , 
                Japan
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Madrid"
                  src="/images/madrid.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Madrid
                , 
                Spain
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Rome"
                  src="/images/rome.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Rome
                , 
                Italy
              </p>
            </div>
            <div
              class="card "
            >
              <div
                class="wrapper-image-card"
              >
                <img
                  alt="Jerusalem"
                  src="/images/jerusalem.webp"
                />
              </div>
              <p
                class="card-title"
              >
                Jerusalem
                , 
                Israel
              </p>
            </div>
          </div>
        </section>
      </section>
      <aside>
        <header>
          <ul>
            <li>
              <label
                class="toggle-wrapper"
                for="toggle"
                tabindex="0"
              >
                <div
                  class="toggle disabled"
                >
                  <input
                    id="toggle"
                    name="toggle"
                    type="checkbox"
                  />
                  <div
                    class="toggle-selector"
                  >
                    <div
                      class="wrapper-icons"
                    >
                      <svg
                        class="sun"
                        fill="#E9E9E9"
                        height="24"
                        viewBox="0 0 24 24"
                        width="24"
                        xmlns="http://www.w3.org/2000/svg"
                      >
                        <path
                          d="M6.995 12C6.995 14.761 9.241 17.007 12.002 17.007C14.763 17.007 17.009 14.761 17.009 12C17.009 9.239 14.763 6.993 12.002 6.993C9.241 6.993 6.995 9.239 6.995 12ZM11 19H13V22H11V19ZM11 2H13V5H11V2ZM2 11H5V13H2V11ZM19 11H22V13H19V11Z"
                        />
                        <path
                          d="M5.63702 19.778L4.22302 18.364L6.34402 16.243L7.75802 17.657L5.63702 19.778Z"
                        />
                        <path
                          d="M16.242 6.34405L18.364 4.22205L19.778 5.63605L17.656 7.75805L16.242 6.34405Z"
                        />
                        <path
                          d="M6.34402 7.75902L4.22302 5.63702L5.63802 4.22302L7.75802 6.34502L6.34402 7.75902Z"
                        />
                        <path
                          d="M19.778 18.3639L18.364 19.7779L16.242 17.6559L17.656 16.2419L19.778 18.3639Z"
                        />
                      </svg>
                    </div>
                  </div>
                </div>
              </label>
            </li>
            <li>
              <a>
                Notifications 
                <span
                  class="alerts"
                >
                  3
                </span>
              </a>
            </li>
            <li>
              <a>
                Places
              </a>
            </li>
          </ul>
          <div
            class="user-avatar"
          />
        </header>
      </aside>
    </article>
    <footer>
      <p>
        Made with ❤️ by 
        <a
          href="https://palmeida.netlify.app"
        >
          Philipe Almeida
        </a>
         
      </p>
    </footer>
  </main>
</div>
`;
